environment_secrets.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2017 Rohith All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package marathon
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// Alias aliases the Application struct so that it will be marshaled/unmarshaled automatically
+type Alias Application
+
+// TmpEnvSecret holds the secret values deserialized from the environment variables field
+type TmpEnvSecret struct {
+	Secret string `json:"secret,omitempty"`
+}
+
+// TmpSecret holds the deserialized secrets field in a Marathon application configuration
+type TmpSecret struct {
+	Source string `json:"source,omitempty"`
+}
+
+// UnmarshalJSON unmarshals the given Application JSON into an environment variables and secrets.
+func (app *Application) UnmarshalJSON(b []byte) error {
+	aux := &struct {
+		*Alias
+		Env     map[string]interface{} `json:"env,omitempty"`
+		Secrets map[string]TmpSecret   `json:"secrets,omitempty"`
+	}{
+		Alias: (*Alias)(app),
+	}
+	if err := json.Unmarshal(b, aux); err != nil {
+		return err
+	}
+	env := make(map[string]string)
+	secrets := make(map[string]Secret)
+
+	if aux.Env != nil {
+		for k, v := range aux.Env {
+			if s, ok := v.(string); ok {
+				env[k] = s
+				continue
+			}
+			tmp, err := json.Marshal(v)
```

Comment:
We won't need reflection but a bit of type-asserting and type-casting. For the sake of comparability, I put together a working implementation and pushed it to my personal repo. Have a look at [this diff](https://github.com/SayWhat1/go-marathon/compare/secrets-support...timoreimann:299-secrets-support).

I'm on the fence on which approach I should find better (or rather less obscure): The extra marshal/unmarshal seems non-obvious (and possibly slower) while the other one is fairly low-level and far from pretty. WDYT? I can probably live with either.